# P2 - Perceptron learning rule

In [10]:
from perceptron import Perceptron, loss
import random

### train een perceptron tot een AND-poort

In [11]:
train_AND = Perceptron([random.random(), random.random()], random.random())# make a perceptron with random weights and bias between 0 and 1

AND = Perceptron([0.5, 0.5], -1) # a perceptron that simulates a AND gate te get the right target values.

iterations = 100 # how many times the perceptrons trains.

while iterations > 0:
    inputs = [random.randint(0, 1) for _ in range(len(train_AND.weights))] # make random input for perceptron.
    target = AND.activation(inputs) # get target value for random input.
    
    train_AND.update(inputs, target) # train
    
    iterations -= 1
    
print(train_AND)

<Perceptron weights: [0.1988124228862785, 0.2849661211777159] en bias: -0.413239488991472>


de weights zijn ongeveer: 0.2 en 0.28 en de bias: -0.4.
met deze weights en bias heb je een AND-poort.

dubbelcheck met de loss functie.

In [18]:
preds = []
targs = []

iterations = 1000 # how many times the perceptrons trains.

"""make a list with predictions and the target value"""
while iterations > 0:
    inputs = [random.randint(0, 1) for _ in range(len(train_AND.weights))] # make random input for perceptron.
    preds.append(train_AND.activation(inputs)) # put prediction in prediction list
    
    targs.append(AND.activation(inputs)) # put target in target list
    
    iterations -= 1

"""get total loss of all the iterations""" 
loss(preds, targs)

0.0

de mean squared error is 0 dus hij klopt.

### train een perceptron tot een XOR-poort

In [43]:
train_XOR = Perceptron([random.random(), random.random()], random.random())# make a perceptron with random weights and bias between 0 and 1

iterations = 10000 # how many times the perceptrons trains.

while iterations > 0:
    inputs = [random.randint(0, 1) for _ in range(len(train_AND.weights))] # make random input for perceptron.
    target = inputs[0]^inputs[1] # get target value for random input.
    
    train_XOR.update(inputs, target) # train
    
    iterations -= 1
    
print(train_XOR)

<Perceptron weights: [0.09437661740721265, 0.016937175932754484] en bias: -0.0014863021685337652>


de weights zijn ongeveer: 0.1 en 0.01 en de bias: -0.001.
met deze weights en bias heb je geen XOR-poort. je kan een XOR-poort ook niet maken met een perceptron.

dubbelcheck met de loss functie.

In [45]:
preds = []
targs = []

iterations = 1000 # how many times the perceptrons trains.

"""make a list with predictions and the target value"""
while iterations > 0:
    inputs = [random.randint(0, 1) for _ in range(len(train_XOR.weights))] # make random input for perceptron.
    preds.append(train_XOR.activation(inputs)) # put prediction in prediction list
    
    targs.append(inputs[0]^inputs[1]) # put target in target list
    
    iterations -= 1

"""get total loss of all the iterations""" 
loss(preds, targs)

0.234

de laagste mean squared error die ik kon krijgen is 0.234.

## train een preceptron op de Iris-dataset

In [57]:
from sklearn.datasets import load_iris

random.seed('1758786') # student nummer as seed.

iris_data = load_iris()
target = list(iris_data.target)
data = list(iris_data.data)

### train de perceptron op het classificeren van de types Setosa en Versicolour

In [90]:
set_ver_data = data[:-50] # get only setosa and versicolour data.
set_ver_target = target[:-50]

"""make iris perceptron with radom weights and bias"""
iris1 = Perceptron([random.random(), random.random(), random.random(), random.random()], random.random())

"""train perceptron for every iris in dataset"""
for i in range(len(set_ver_data)):
    iris1.update(set_ver_data[i], set_ver_target[i]) # train

print(iris1)

<Perceptron weights: [-0.21211831941397824, -0.29109048163645085, 0.5622547032667522, 0.09704799944100864] en bias: 0.11274352433111418>


de weights zijn ongeveer: -0.21, -0.29, 0.56, 0.097 en de bias: 0.11.

accuracy met de loss functie.

In [91]:
preds = []

"""make a list with predictions"""
for i in set_ver_data:
    preds.append(iris1.activation(i)) # put prediction in prediction list
    
    iterations -= 1

"""get total loss of all the iterations""" 
loss(preds, set_ver_target)

0.0

### train de perceptron op het classificeren van de types Versicolour en Verginica

In [149]:
ver_ver_data = data[:100] # get only Versicolour en Verginica data.
ver_ver_target = target[:100]

"""make iris perceptron with radom weights and bias"""
iris2 = Perceptron([random.random(), random.random(), random.random(), random.random()], random.random())

"""train perceptron for every iris in dataset"""
for i in range(len(ver_ver_data)):
    iris2.update(ver_ver_data[i], ver_ver_target[i]) # train

print(iris2)

<Perceptron weights: [-0.06514374266736894, -0.4749663714961107, 0.23969593873015807, 0.6258756228188677] en bias: 0.21849595972973876>


de weights zijn ongeveer: -0.18, 0.38, 0.38, 0.6 en de bias: 0.62.

accuracy met de loss functie.

In [150]:
preds = []

"""make a list with predictions"""
for i in ver_ver_data:
    preds.append(iris2.activation(i)) # put prediction in prediction list
    
    iterations -= 1

"""get total loss of all the iterations""" 
loss(preds, ver_ver_target)

0.0

de mean squared error is altijd 0 of 0.5

In [151]:
for i in range(len(preds)):
    print(preds[i])
    print(ver_ver_target[i])

0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1


de perceptron voorspelt soms altijd dezelfde waarde. dit zorgd ervoor dat de mean squared error dan 0.5 is omdat de voorspellende waardes 0 of 1 zijn.

kijken of dit probleem zich oplost door de perceptron meerdere keren trainen op de dataset. 

In [164]:
"""make iris perceptron with radom weights and bias"""
iris3 = Perceptron([random.random(), random.random(), random.random(), random.random()], random.random())

epochs = 5
"""train perceptron for every iris in dataset"""
while epochs > 0:
    for i in range(len(ver_ver_data)):
        iris3.update(ver_ver_data[i], ver_ver_target[i]) # train
    epochs -= 1

print(iris3)

<Perceptron weights: [-0.05237053961268051, -0.48467192298636463, 0.30007514815621283, 0.1544878898840285] en bias: 0.6766862211648206>


In [165]:
preds = []

"""make a list with predictions"""
for i in ver_ver_data:
    preds.append(iris3.activation(i)) # put prediction in prediction list
    
    iterations -= 1

"""get total loss of all the iterations""" 
loss(preds, ver_ver_target)

0.0

met 5 epochs verspelt de perceptron niet meer steeds een dezelfde waarde. en voorspelt de perceptron met meer epcohs accurater